LiveVisitsFromLogs.php

```diff
@@ -0,0 +1,204 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\VisitorGenerator;
+
+
+use Piwik\Container\StaticContainer;
+use Piwik\Date;
+use Piwik\Http;
+use Piwik\Plugins\VisitorGenerator\Faker\Request;
+use Piwik\Plugins\VisitorGenerator\Generator\VisitsFromLogs;
+use Piwik\Site;
+use Psr\Log\LoggerInterface;
+use SplFileObject;
+
+class LiveVisitsFromLogs
+{
+    const MAX_LOG_LINES_TO_SKIP = 10;
+
+    /**
+     * @var Site
+     */
+    private $site;
+
+    /**
+     * @var int
+     */
+    private $maxConcurrentActions;
+
+    /**
+     * @var string[]
+     */
+    private $logFiles;
+
+    /**
+     * @var VisitsFromLogs
+     */
+    private $visitsFromLogs;
+
+    /**
+     * @var int
+     */
+    private $currentLine = 0;
+
+    /**
+     * @var int
+     */
+    private $currentFile = 0;
+
+    /**
+     * @var SplFileObject
+     */
+    private $currentFileHandle = null;
+
+    public function __construct($piwikUrl, Site $site, $maxConcurrentActions, array $logFiles, $initialFile = null, $initialLine = 0)
+    {
+        $this->site = $site;
+        $this->maxConcurrentActions = $maxConcurrentActions;
+        $this->logFiles = $this->filterEmptyFiles($logFiles);
+        $this->visitsFromLogs = new VisitsFromLogs($piwikUrl);
+
+        if (!empty($initialFile)) {
+            $this->currentFile = (int) array_search($initialFile, $logFiles);
+        }
+
+        $this->currentLine = $initialLine;
+
+        $this->openFile();
+        $this->currentFileHandle->seek($this->currentLine);
+    }
+
+    public function track()
+    {
+        $logger = StaticContainer::get(LoggerInterface::class);
+
+        $numberOfActions = rand(1, $this->maxConcurrentActions);
+        $acceptLanguages = Request::getAcceptLanguages();
+
+        $this->skipRandomNumberOfLogLines();
```

Comment:
I think ideally we would not process requests randomly but pretty much replay the logs as they are defined. Eg when it is the 5th of a month at 13:40 and we tell the command to run for 5 minutes, then we would ideally for the next 5 minutes replay all logs from the 5th of a month between 13:40 and 13:45 as if it was real-time traffic basically.  The 5th of a month assumes we have logs for eg 31 days in there.

If we start tracking requests randomly, the tracked information may not look as realistic anymore? We also need to make sure to have all requests of a visit next to each other I would say etc.

To not show the same evolution screens every month, we could maybe have 1.5-2 months worth of log data and on every months or days also track data from other months. This "randomness" would need to be consistent though (a contradiction in terms). For example if we have another 30days of data in the logs which are usually not replayed (43200 minutes), and assuming half a year has 183days, every day we could replay additional 236minutes/4 hours of the other logs every day. So on day 1 of the year, we replay all logs from the first day (from the first 31 days of logs), plus the first 4 hours of the other month (from the second 31 days of logs). On day 2 of the year, we replay all logs from the 2nd day (from the first 31 days of logs), plus the hours 4 to 8 from the other month (from the second 31 days of logs). 

I know I'm explaining it not well and it doesn't even have to be that solution. I'm just trying to avoid too much "randomness" as we ideally replay as much as things happened in the logs. 

In case we did want to create much more traffic than the logs have, then the suggested solution wouldn't work. Also I haven't really thought it through, the suggested solution might not really work. Just trying to say it would be great to replay logs in order as they are defined. 

Maybe it is not a problem to replay them randomly by skipping some lines and we would just need to test what it looks like. Ideally we prevent though a visit starting with some random action such as a download, purchase, ... and often a visit should start on the homepage I suppose etc.

Reply:
Perhaps it would help to determine exactly what we can allow to be random.

I guess the start of a visit shouldn't be random because we don't want half a real visit in there? What about the start time of a visit? For example, if a visit from 10 AM was instead fully replayed at 10 PM, would that make the data less realistic or more varied?

Another idea might be to just replay the logs as usual, playing the current visits for the current day of month/time as you said, but then replace the logs that are used every month. So for February, we'd have January's logs there, and would replay that. And for March, we'd replace it w/ February's logs. This assumes we can continuously collect these logs w/o issue.

What do you think @tsteur ?

Reply:
> I guess the start of a visit shouldn't be random because we don't want half a real visit in there? What about the start time of a visit? For example, if a visit from 10 AM was instead fully replayed at 10 PM, would that make the data less realistic or more varied?

That wouldn't make less realistic data. Problem is just in looking through the logs and checking for each visit when the logs started and making sure to include all of them. If that's not too complicated, we could likely also replay more traffic randomly that way.

Otherwise maybe also a more simple solution would do as we won't have ongoing new logs. Say we had 2 months of data, we could just track every 2 months the same data, at least in V1 for simplicity? The graphs would repeat every 2 months but I reckon it would be kind of fine as in a demo you don't look too far back anyway etc.

Reply:
That sounds like a good start, and (IMO) it would be much better to have something up and working that we can iterate on, than to spend a lot of time hoping we find the perfect solution. I'll go w/ that and let's see if we can set everything up.